### PR TITLE
Implement SanctoraleWriter

### LINKED ISSUE #60: SanctoraleDumper
It would be handy to have a class doing the opposite of what `SanctoraleLoader` does - serializing contents of a `Sanctorale` instance to the [standard data file format](https://github.com/igneus/calendarium-romanum/tree/master/data).

### PR DESCRIPTION
**Pull request by mkasberg**
`SanctoraleLoader` is a class that already exists to read a txt file
into a `Sanctorale` instance. It would be nice to also have a
`SanctoraleWriter` class to do this operation in reverse, writing
`Sanctorale` data to a txt file in the same format.

The `SanctoraleWriter` is probably less complicated than the loader
because we don't need to do any regex or other pattern matching. We
simply iterate the celebrations in the `Sanctorale` and write a line for
each.

There was some room for judgement here because the file format allows
for a couple different variations of the way months or ranks could be
serialized. I opted to serialize using the preferences that
`universal-en.txt` was already using, which also seemed to make the file
the most readable. (E.g. We serialize month headings rather than writing
the month on every line). An added benefit is that we can deserialize
and then serialize `universal-en.txt` and compare the results. I've done
this, and the results are identical, except for the comment in the YAML
frontmatter (which isn't stored in the Sanctorale).

This fixes #60.

### CODE REVIEW

**Comment by mkasberg** [comment]
Let me know what you think! Happy to work with you on any changes needed :slightly_smiling_face:

**Review by igneus — CHANGES_REQUESTED** [review]
Thank you very much, looks good overall.

I've added a few comments. Apart of addressing them please make sure, that tests pass also on Ruby 2.0, which is (ancient, but) still supported by calendarium-romanum https://travis-ci.org/github/igneus/calendarium-romanum/jobs/732125260

**Comment by igneus** [comment]
One more thing: there should be a simple integration test making sure that (non-empty) output of `SanctoraleWriter` is accepted by `SanctoraleLoader`.

File: lib/calendarium-romanum/sanctorale_writer.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    # @param dest [String, File, #<<]
+    #   object to populate. If not provided, a new {String}
+    #   instance will be created and returned
+    # @return [String]
+    def write(src, dest = nil)
+      dest ||= String.new
+
+      # Write metadata to YAML if present
+      unless src.metadata.nil? || src.metadata.empty?
+        dest << src.metadata.transform_keys(&:to_s).to_yaml
```
[the remark this answers is not in the document]
Comment (igneus) [reply lib/calendarium-romanum/sanctorale_writer.rb:57]: Looking at this for the second time, I think the call to `#transform_keys` is superfluous and can be simply deleted. It's not `SanctoraleWriter`'s responsibility to care about contents of the metadata and result of their serialization.

File: spec/sanctorale_writer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          expect(result).to include(' R ')
+        end
+      end
+    end
+  end
+
+  describe 'exceptional cases' do
+    it 'raises an exception if no date is provided' do
+      c = CR::Celebration.new
+      expect { d.send(:celebration_line, c) }.to raise_exception(NoMethodError)
```
Comment (mkasberg) [inline spec/sanctorale_writer_spec.rb:101]: While trying to remove unnecessary fields, I found a small issue here in the case where no date is present. I think we can handle it better, but I'm not sure what the best approach is. Should a date be a required parameter (i.e. should the `Celebration` constructor raise an exception if `date` is not present)? Should we raise an error during serialization if we try to serialize a `Celebration` without a date?

Comment (igneus) [reply spec/sanctorale_writer_spec.rb:101]: Ouch, in my code review I missed that in `#celebration_line` you use `Celebration#date` as the date of the celebration to be serialized. That's not correct. While it may work in "usual" cases, values of `Celebration#date` are not part of `Sanctorale`'s contract. `Sanctorale` does not promise that all `Celebration`s it returns have non-nil `#date`, nor that the `#date` (if non-nil) is the same as the date the `Celebration` happens to occur on in this particular `Sanctorale` instance.

Please change `#celebration_line` to serialize the date used as key in `Sanctorale` (and yielded by `#each_day`) instead.

Comment (mkasberg) [reply spec/sanctorale_writer_spec.rb]: Done. Lucky for me, this was a pretty simple adjustment :slightly_smiling_face:

**Comment by igneus** [comment]
Thank you very much, merging.
